### PR TITLE
chore: add mutex on KeyringSigner for QueryAccount

### PR DESCRIPTION
**Pull request by albttx**
## Description

closes: #143

### CODE REVIEW

File: x/payment/types/builder.go [hunk 1]
```diff
@@ -44,6 +47,8 @@ func (k *KeyringSigner) QueryAccountNumber(ctx context.Context, conn *grpc.Clien
 		return err
 	}
 
+	k.Lock()
+	defer k.Unlock()
```
Comment (Wondertan) [inline x/payment/types/builder.go:51]: We should not lock the mutex for the entire network request inside `QueryAccount`. Instead, the lock should be applied to the read/write of shared memory.

**Review by Wondertan — COMMENTED** [review]
Hello @albttx. Thanks for PR. A question: why do you think we need a mutex there? AFAIU, we don't use the method concurrently. Although, it does not hurt to have it. Also, the synchronization should likely be consistent with all other methods altering private fields.

**Review by evan-forbes — CHANGES_REQUESTED** [review]
Thanks for doing this @albttx! We appreciate this a lot.

I agree with @Wondertan, we only need to enforce a lock when reading and writing. Therefore, we simply have to lock after making the request instead of before, and then another write lock here
https://github.com/celestiaorg/lazyledger-app/blob/b6eafb59e4141fdd8891c964c261511a5fc01678/x/payment/types/builder.go#L134-L142

we also have to create a read lock when we access the sequence and account numbers to create a transaction.

https://github.com/celestiaorg/lazyledger-app/blob/b6eafb59e4141fdd8891c964c261511a5fc01678/x/payment/types/builder.go#L77-L107

thanks again, and definitely let me know if you have any further questions :slightly_smiling_face:

**Comment by evan-forbes** [comment]
also, @Wondertan 
> A question: why do you think we need a mutex there? AFAIU, we don't use the method concurrently. Although, it does not hurt to have it.

As you stated, the only reason to include it is that it doesn't hurt to have it. While we don't use in a fashion that requires it, it is meant to code used as a library for clients, so I think the extra protection is warranted and should have been added when this code was initially introduced.

**Comment by albttx** [comment]
Hello @evan-forbes i've just updated my PR, thanks for the explanations

So now all `KeyringSigner.{accountNumber,sequence}` are now mutex protected.

File: x/payment/types/builder.go [hunk 2]
```diff
@@ -61,7 +67,12 @@ func (k KeyringSigner) NewTxBuilder() sdkclient.TxBuilder {
 // BuildSignedTx creates and signs a sdk.Tx that contains the provided message. The interal
 // account number must be set by calling k.QueryAccountNumber or by manually setting it via
 // k.SetAccountNumber for the built transactions to be valid.
-func (k KeyringSigner) BuildSignedTx(builder sdkclient.TxBuilder, msg sdktypes.Msg) (authsigning.Tx, error) {
+func (k *KeyringSigner) BuildSignedTx(builder sdkclient.TxBuilder, msg sdktypes.Msg) (authsigning.Tx, error) {
+	k.RLock()
+	accountNumber := k.accountNumber
+	sequence := k.sequence
+	k.RUnlock()
```
Comment (evan-forbes) [inline x/payment/types/builder.go:74]: :+1:

**Review by evan-forbes — APPROVED** [review]
perfect, thanks! :purple_heart: :sparkles:

**Comment by albttx** [comment]
Hello, sorry i just noticed that some tests were broken, the `sync.RWMutex` was a pointer, wich is not required, just the methods on `KeyringSigner` need to be

**Comment by evan-forbes** [comment]
yeah good call, let's figure out all the linting and tests before merging

**Comment by albttx** [comment]
updated, could you restart the linter please :)

**Review by evan-forbes — APPROVED** [review]
super dope, @albttx 

thanks for fixing everything!

**Pull request merged**